### PR TITLE
[WebNN EP] Support conv1d by reshaping with prepended 1's

### PR DESCRIPTION
**Pull request by Honry**
WebNN only supports 4-D inputs for conv2d and convTranspose2d, this PR supports 3-D inputs (i.e. conv1d) by prepending a 1 size dimension and several reshape operations.

### CODE REVIEW

**Comment by Honry** [comment]
@fdwr, @guschmue, PTAL when you're back. Thanks!

And enjoy your holiday~

**Comment by fdwr** [comment]
> prepending a 1 size dimension and several reshape operations

🤔 This will work. Though, do you think it makes more sense to just generalize WebNN's conv operator, and do it there? I'd hate for each calling ML library to have to repeat this chunk of reshaping logic. For the DML backend, everything just gets ferried along (as it's Conv accepts 3D to 5D inputs, or rather 1D to 3D spatial dimensions). For XNNPack, it could branch to the respective conv implementation depending on the rank, or extend the conv like you did here.

**Comment by Honry** [comment]
> > prepending a 1 size dimension and several reshape operations
> 
> 🤔 This will work. Though, do you think it makes more sense to just generalize WebNN's conv operator, and do it there? I'd hate for each calling ML library to have to repeat this chunk of reshaping logic. For the DML backend, everything just gets ferried along as-is (it's Conv accepts 3D to 5D inputs, or rather 1D to 3D spatial dimensions). For XNNPack, it could branch to the respective conv implementation depending on the input rank, or extend the conv attributes like you did here.

Can't agree more. It's painful for me to integrate the conv1d in this complex file. 

We are discussing the necessity of integrating WebNN EP into JS EP, you know it should be a part of JS EP, and it would greatly benefit from its extensibility and enable us to support the WebNN Async API. We will talk this with you and ONNXRuntime guys after the holiday.

So I'd suggest leave this change as is to reduce the duplicate effort. We can do it in near future, maybe in the JS EP. :)

File: onnxruntime/core/providers/webnn/builders/helper.h [hunk 1]
```diff
@@ -51,6 +51,15 @@ std::string GetShapeString(std::vector<T>& shape) {
...
 }
 
+inline std::vector<uint32_t> GetVecUint32FromVecInt64(const std::vector<int64_t>& int64_vec) {
+  std::vector<uint32_t> uint32_vec;
+  uint32_vec.reserve(int64_vec.size());
+  std::transform(int64_vec.begin(), int64_vec.end(),
+                 std::back_inserter(uint32_vec),
+                 [](int64_t val) -> uint32_t { return SafeInt<uint32_t>(val); });
+  return uint32_vec;
+}
```
Comment (fdwr) [inline onnxruntime/core/providers/webnn/builders/helper.h:61]: Yay, happy to see this function after seeing `transform` repeated so many times.

File: onnxruntime/core/providers/webnn/builders/impl/conv_op_builder.cc [hunk 2]
```diff
@@ -42,72 +42,65 @@ void ConvOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const Nod
...
     // in options.outputPadding are ignored.
-    std::vector<int32_t> dim;
-    std::vector<int32_t> output_padding{0, 0};
+    std::vector<int64_t> dims;
+    std::vector<int64_t> output_padding{0, 0};
     if (helper.HasAttr("output_shape")) {
-      // Default value of 'output_shape' will be ignore as we already check if
-      // it's existed.
-      dim = helper.Get("output_shape", std::vector<int32_t>{-1, -1});
+      // Default value of 'output_shape' will be ignore as we already check if it's existed.
```
Comment (fdwr) [inline onnxruntime/core/providers/webnn/builders/impl/conv_op_builder.cc:72]: `ignored`

`if it existed.`

File: onnxruntime/core/providers/webnn/builders/impl/conv_op_builder.cc [hunk 3]
```diff
@@ -42,72 +42,65 @@ void ConvOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const Nod
...
-      } else if (dim.size() == 4) {
-        output_shape = {dim[2], dim[3]};
+      std::vector<int64_t> output_shape;
+      if (dims.size() == 1 && is_conv1d) {  // ConvTranspose 1d
+        output_shape = {dims[0], 1};
+      } else if (dims.size() == 2 && !is_conv1d) {
+        output_shape = dims;
+      } else if (dims.size() == 3 && is_conv1d) {  // ConvTranspose 1d
+        output_shape = {dims[2], 1};
+      } else if (dims.size() == 4 && !is_conv1d) {
```
Comment (fdwr) [inline onnxruntime/core/providers/webnn/builders/impl/conv_op_builder.cc:82]: When will the ConvTranspose `output_shape` be 4? The batch and channel dimensions are excluded from `output_shape`, meaning it should only include the spatial dimensions, and WebNN only supports 2 spatial dimensions (DML only supports 1-3 spatial dimensions). 

https://onnx.ai/onnx/operators/onnx__ConvTranspose.html

File: onnxruntime/core/providers/webnn/builders/impl/conv_op_builder.cc [hunk 4]
```diff
@@ -250,38 +253,84 @@ Status ConvOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, const N
...
+    std::vector<uint32_t> new_shape = GetVecUint32FromVecInt64(input_shape);
+    input = model_builder.GetBuilder().call<emscripten::val>("reshape", input, emscripten::val::array(new_shape));
+
+    weight_shape.push_back(1);
+    if (strides.size() == 1) {
+      strides.push_back(1);
+    }
+    if (dilations.size() == 1) {
+      dilations.push_back(1);
+    }
```
Comment (fdwr) [inline onnxruntime/core/providers/webnn/builders/impl/conv_op_builder.cc:288]: How about?

```
      weight_shape.resize(4, 1);  // Ensure 4D by appending 1's if needed.
      strides.resize(2, 1);       // Ensure 2D by appending 1's if needed.
      dilations.resize(2, 1);     // Ensure 2D by appending 1's if needed.
```

**Review by fdwr — COMMENTED** [review]
Some thoughts. Thanks for adding 1D.

File: onnxruntime/core/providers/webnn/builders/impl/conv_op_builder.cc (hunk 4 above)
Comment (Honry) [reply onnxruntime/core/providers/webnn/builders/impl/conv_op_builder.cc]: Fixed.

**Review by Honry — COMMENTED** [review]
@fdwr, thanks for your comments, I've addressed them and rebased the code, PTAL again, thanks!

**Review by guschmue — APPROVED** [review]

**Review by fdwr — DISMISSED** [review]

**Comment by Honry** [comment]
@fdwr, I just fix the lint error, pls. help retrigger the CI. thanks!

**Review by fdwr — APPROVED** [review]
Sheesh, did the linter complain that the comments weren't aligned? Reapproving.

**Comment by Honry** [comment]
> Sheesh, did the linter complain that the comments weren't aligned? Reapproving.

Thanks, it doesn't specific the reason, I just format it again. :p

**Pull request merged**